compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -721,7 +721,7 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
         case fdef: js.FieldDef =>
           implicit val pos = fdef.pos
           val select = fdef.name match {
-            case ident: js.Ident          => js.JSPrivateSelect(selfRef, ident)
+            case ident: js.Ident          => js.JSPrivateSelect(selfRef, encodeClassRef(sym), ident)
```

Comment:
Seeing the code with the validation makes me wonder if this line here is not a potential problem for binary compatibility: We are basically forcing the linker to emit a private JS field as:

- Something static somewhere (based on the `Ident`).
- Assignments (where the initial assignment is a-priori indistinguishable from any other assignment).

Contrast this to non-local classes where we have the ability to intercept the constructor function (and we do to some extent, especially for ES5 for prototype chaining).

Are we OK with this restriction on how JS private fields are "created"?

Reply:
I don't think this is problematic. Local classes are, by spec, reverse engineered to behave as if the class had never existed in the first place, and the object had been created from the parent class all along. I believe the restriction you're talking about is the semantically correct thing to do.

Reply:
Hum, but by that definition the parent class cannot be abstract and the implementation cannot access the private field (granted, we could specify `JSPrivate` to mean privately scoped, not private as in class private, which is probably the right thing to do anyways if we want IRCheck to work post opt ever).

Also, this definition does not explain how it is justifiable that the fields end up as static fields on the corresponding abstract JS type (which IIUC, is merely named the same by happenstance and used as a holder for static methods, and not instantiatable).

Writing this, I realize:
- The parent not being abstract is a real requirement so this is WAI.
- If we define JS private fields as "privately scoped" we should maybe put them in the static namespace all the time and merely have them mean: this field needs to be usable. (this might warrant a different tree, but I'm unsure about that).

Reply:
Our private do mean "private to Scala.js" (not visible to JavaScript). Indeed they don't mean private per class.

> the fields end up as static fields on the corresponding abstract JS type

They're actually instance fields (not static) in the abstract JS type. And yes it basically means "this field must be usable".

I'm not 100% happy with the situation, but I don't have a better idea at the moment. Eventually I'd like not to use private fields at all in anonymous classes, and instead generate a *captured* state object to hold the private fields of an anonymous class.

Reply:
> They're actually instance fields (not static) in the abstract JS type. 

Right, the line `staticMembers += ...` tripped me off (we should probably rename that accumulator).

> Eventually I'd like not to use private fields at all in anonymous classes, and instead generate a captured state object to hold the private fields of an anonymous class.

That sounds reasonable. Do you think this is feasible before 1.0 so we do not have to allow the strange field capability?

Otherwise, we basically have to allow to set any JS private field on any JS value (and it should work). It seems that only allowing to set them on values of the type they refer to is more reasonable (by allowing I mean runtime, not compile time semantics).

Another way to think about it: With the current semantics we couldn't layout a class at definition time (i.e. post capture). That seems not a good thing to commit to (I'm ok with it as an intermediate step).

Reply:
> Do you think this is feasible before 1.0 so we do not have to allow the strange field capability?

I hope so, yes. I would like to be able to remove the instance fields on abstract JS types, and as you say, it would be nice to specify that private JS field selection is undefined behavior if the receiver is not an instance of the class that declares the field. (We could even check that at run-time in fastOpt, actually.)

Reply:
I have submitted #3777 for this.